src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketPullRequestDiscoveryTrait.java

```diff
@@ -0,0 +1,140 @@
+package com.atlassian.bitbucket.jenkins.internal.scm.trait;
+
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketClientFactoryProvider;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketPluginConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
+import com.atlassian.bitbucket.jenkins.internal.scm.*;
+import com.cloudbees.plugins.credentials.Credentials;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import jenkins.plugins.git.GitSCMBuilder;
+import jenkins.plugins.git.MergeWithGitSCMExtension;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadCategory;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import jenkins.scm.impl.ChangeRequestSCMHeadCategory;
+import jenkins.scm.impl.trait.Discovery;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class BitbucketPullRequestDiscoveryTrait extends BitbucketSCMSourceTrait {
+
+    @DataBoundConstructor
+    public BitbucketPullRequestDiscoveryTrait() {
+    }
+
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        if (context instanceof BitbucketSCMSourceContext) {
+            BitbucketSCMSourceContext bitbucketContext = (BitbucketSCMSourceContext) context;
+            BitbucketSCMRepository repository = bitbucketContext.getRepository();
+            bitbucketContext.withDiscoveryHandler(
```

Comment:
All this trait does is register an implementation of [BitbucketSCMHeadDiscoveryHandler](https://github.com/jaguilar-atl/atlassian-bitbucket-server-integration-plugin/pull/2/files#diff-a2a71512cc815d2b3e2d0b398e896982bd0fe7ffec9930b7e352aebe008878b8) which gets called [in the retrieve method here](https://github.com/jaguilar-atl/atlassian-bitbucket-server-integration-plugin/pull/2/files#diff-ed9f29c11fc5b996df33a07d0643914c98ac75e7a28a3f90335963929725c5eeR333).

This means that it will be the concern of the discovery trait to define what a head is and how it will be fetched. The retrieve method only sees that it is there and needs to be processed.